tower-http/src/set_response_header.rs

```diff
@@ -0,0 +1,457 @@
+//! use hyper::Body;
+//! use http_body::Body as _; // for `Body::size_hint`
+//!
+//! # #[tokio::main]
+//! # async fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! # let render_html = tower::service_fn(|request: Request<Body>| async move {
+//! #     Ok::<_, std::convert::Infallible>(Response::new(Body::from("1234567890")))
+//! # });
+//! #
+//! let mut svc = ServiceBuilder::new()
+//!     .layer(
+//!         // Layer that sets `Content-Length` if the body has a known size.
+//!         // Bodies with streaming responses wont have a known size.
+//!         SetResponseHeaderLayer::new(
+//!             http::header::CONTENT_LENGTH,
+//!             |response: &Response<Body>| {
+//!                 if let Some(size) = response.body().size_hint().exact() {
+//!                     // If the response body has a known size, returning `Some` will
+//!                     // set the `Content-Length` header to that value.
+//!                     Some(HeaderValue::from_str(&size.to_string()).unwrap())
+//!                 } else {
+//!                     // If the response body doesn't have a known size, return `None`
+//!                     // to skip setting the header on this response.
+//!                     None
+//!                 }
+//!             }
+//!         )
+//!     )
+//!     .service(render_html);
+//!
+//! let request = Request::new(Body::empty());
+//!
+//! let response = svc.ready_and().await?.call(request).await?;
+//!
+//! assert_eq!(response.headers()["content-length"], "10");
+//! #
+//! # Ok(())
+//! # }
+//! ```
+
+use futures_util::ready;
+use http::{header::HeaderName, HeaderValue, Response};
+use pin_project::pin_project;
+use std::{
+    fmt,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use std::{future::Future, marker::PhantomData};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Layer that applies [`SetResponseHeader`] which adds a response header.
+///
+/// See [`SetResponseHeader`] for more details.
+pub struct SetResponseHeaderLayer<M, T> {
+    header_name: HeaderName,
+    make: M,
+    mode: InsertHeaderMode,
+    _marker: PhantomData<fn() -> T>,
+}
+
+impl<M, T> fmt::Debug for SetResponseHeaderLayer<M, T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("SetResponseHeaderLayer")
+            .field("header_name", &self.header_name)
+            .field("mode", &self.mode)
+            .field("make", &std::any::type_name::<M>())
+            .finish()
+    }
+}
+
+impl<M, T> SetResponseHeaderLayer<M, T> {
+    /// Create a new [`SetResponseHeaderLayer`].
+    pub fn new(header_name: HeaderName, make: M) -> Self
+    where
+        M: MakeHeaderValue<T>,
+    {
+        Self {
+            make,
+            header_name,
+            mode: InsertHeaderMode::OverrideExisting,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Set which mode to use when inserting the header.
```

Comment:
nit: "sets which mode" isn't terribly descriptive. I might reword this to focus on what it actually configures, rather than just saying the name of the type.

how about something like this:
```suggestion
    /// Configures how existing header values are handled.
    ///
    /// This takes an [`InsertHeaderMode`] which configures the service's
    /// behavior when other values have previously been set for the same
    /// header. The available options are:
    ///
    /// - `InsertHeaderMode::OverrideExisting` (the default): if a previous
    ///   value exists for the same header, it is removed and replaced with
    ///   the new header value.
    /// - `InsertHeaderMode::SkipIfPresent`: if a previous value exists for
    ///   the header, the new value is not inserted.
    /// - `InsertHeaderMode::Append`: the new header is always added,
    ///   preserving any existing values. If previous values exist, the header
    ///   will have multiple values.
```